tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -458,6 +459,81 @@ where
+        if let Some(timings) = extensions.get_mut::<Timings>() {
+            timings.entered = Some(Instant::now());
+        } else {
+            extensions.insert(Timings::new());
+        }
+    }
+
+    fn on_exit(&self, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+        if let Some(timings) = extensions.get_mut::<Timings>() {
+            if let Some(started) = timings.entered.take() {
+                timings.accumulated += started.elapsed().as_nanos() as u64;
+            }
+        }
+    }
+
+    fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+        let span = ctx.span(&id).expect("Span not found, this is a bug");
+        let extensions = span.extensions();
+        if let Some(timing) = extensions.get::<Timings>() {
+            let t = field::display(timing.get_accumulated());
+            let meta = span.metadata();
+            let name = field::display(meta.name());
+            let cs = meta.callsite();
+            let meta = Metadata::new(
+                meta.name(),
+                meta.target(),
+                meta.level().clone(),
+                meta.file(),
+                meta.line(),
+                meta.module_path(),
+                field::FieldSet::new(&[], cs.clone()),
+                Kind::EVENT,
+            );
+            let fs = field::FieldSet::new(&["message", "timing"], cs);
+            let mut it = fs.iter();
+            let v = [
+                (&it.next().unwrap(), Some(&name as &dyn field::Value)),
+                (&it.next().unwrap(), Some(&t as &dyn field::Value)),
+            ];
+            let vs = fs.value_set(&v);
+            let event =
+                Event::new_child_of(id, unsafe { &*(&meta as *const Metadata<'static>) }, &vs);
+
+            let mut buf = String::new();
+            let ctx = self.make_ctx(ctx.clone());
+            if self.fmt_event.format_event(&ctx, &mut buf, &event).is_ok() {
+                let fields = extensions
+                    .get::<FormattedFields<N>>()
+                    .expect("Unable to find FormattedFields in extensions; this is a bug");
+                if !fields.is_empty() {
+                    use std::fmt::Write;
+                    loop {
+                        let last = buf.pop();
+                        match last {
+                            Some(x) if x.is_whitespace() => {}
+                            Some(x) => {
+                                buf.push(x);
+                                break;
+                            }
+                            None => break,
+                        }
+                    }
+                    let _ = writeln!(&mut buf, " {}", fields);
+                }
+                let mut writer = self.make_writer.make_writer();
+                let _ = io::Write::write_all(&mut writer, buf.as_bytes());
+            }
```

Comment:
What do you think about just calling `self.on_event` with the synthesized event, and including the formatted fields of the span in the `message` field? That way we could benefit from the same thread-local buffer that `on_event` formats events into, and avoid having to allocate a string for every span close?

Reply:
Is it somehow possible to transfer the fields from the (close) span into the (synthesised) event? Then that would be the cleanest solution.

Reply:
Hmm, not the field values --- the field values themselves are only available when they are recorded, since they are borrowed from the scope they were recorded in. At this point, we only have their formatted representation.

That said: since we generate the synthesized event as a child of the span that closed, the output of `self.on_event` would already contain the span's fields in the context that's displayed before the event's data. I'm not sure if we _really_ need to also duplicate the fields in the message as well?

Reply:
No, I tried that for quite a while before I understood that that cannot work: the span is closed, so the `.scope()` will not include it when iterating, so the span’s fields are not printed by some existing magic :-(

Reply:
> No, I tried that for quite a while before I understood that that cannot work: the span is closed, so the `.scope()` will not include it when iterating, so the span’s fields are not printed by some existing magic :-(

Hmm, after taking a look at `format_event`, you're right, but that's actually due to a bug in the `format_event` implementation for the default formatters. It looks like they always print the _current_ span's scope as the context, even if the event has an explicitly set parent. We should fix that, and print the event's explicit parent's scope if one is set. Then, using `new_child_of` to create the synthesized event _would_ include that span in the scope (since its data is only removed from the registry _after_ `on_close` completes).

Reply:
Okay, I just merged PR #767,  which fixes the issue where `format_event` did not honor the overridden parent set by `Event::new_child_of`. If we rebase this branch to pick that up, we shouldn't have to bother with the span`s fields at all, since the event will be properly formatted as its child.